docs/replication.md

```diff
@@ -0,0 +1,145 @@
+
+
+
+# Couchdb replication & limitation
+
+Couchdb 2 replication protocol is described [in details here ](http://docs.couchdb.org/en/2.0.0/replication/protocol.html)
+
+
+## Quick summary
+
+1. The replicator figures out what was the last sequence number *lastseqno* which was correctly synced from the source database, using a `_local/:replicationid` document.
+2. The replicator `GET :source/changes?since=lastseqno&limit=batchsize` and identify which docs have changed (and their *open_revs*).
+3. The replicator `POST :target/_revs_diff` to get a list of all revisions of the changed documents the target does not know.
+4. The replicator `GET :source/:docid?open_revs=['rev1', ...]` several times to retrieve the missing documents revisions.
+5. The replicator `POST :target/_bulk_docs` with these documents revisions.
+6. Store the new last sequence number
+
+Repeat 2-5 until there is no more changes.
+
+## Details
+
+- In step 5, the replicator can also attempt to `PUT :target/:docid` if doc are too heavy, but this should not happens in cozy-stack considering there wont be attachment in couchdb.
+- The main difference from couchdb 1.X is in the replication history and the manner to determine and store the last sequence number. **TODO :** actually understand this and how it might increase disk usage if we have a lot of replications.
+- Couchdb `_changes` and by extension replication can be either by polling `?since=xx&limit=xx` can return `[]`, or continuous (SSE / COMET)
+- In couchdb benchmarking, we understood that the number of couch databases only use a bit of disk space but no RAM or CPU **as long as the database is not used**. Having a continuous replication active force us to keep the database file open and will starve RAM & FD usage. Moreover, continuous replication costs a lot (cf. ScienceTeam benchmark). To permit an unlimited number of inactive user on a single cozy-stack process, **the stack should avoid continuous replication from couchdb**.
+- Two way replication is simply two one way replications
+
+
+## Routes used by replication
+
+To be a source of replication, the stack only need to support the following route:
+
+- `GET :source/:docid?open_revs=xxxx` get revisions of a document
+- `GET  :source/_changes` get a list of ID -> New Revision since a given sequence number
+
+To be a target of replication, the stack need to support the following routes:
+
+- `POST :target/_revs_diff` takes a list of ID -> Rev and returns which one are missing.
+- `POST :target/_bulk_docs` create several documents at once
+- `POST :target/_ensure_full_commit` ensure the documents are written to disk. This is useless if couchdb is configured without delayed write (default), but remote couchdb will call, so the stack should return the expected 201
+
+In both case, we need to support
+
+- `PUT    :both/_local/:revdocid` to store the current sequence number.
+
```

Comment:
I think it should be nice to also support [`_all_docs`](http://docs.couchdb.org/en/2.0.0/api/database/bulk-api.html#db-bulk-docs). It is currently used by cozy-mobile to speed up the synchronization. I think pouchdb try to use it for replication, and has some fallbacks when it's not available.